packages/next/src/server/use-cache/use-cache-wrapper.ts

```diff
@@ -376,61 +499,61 @@ export function cache(kind: string, id: string, fn: any) {
-        // stale entry since it would unnecessarily need to shorten the lifetime of the
-        // prerender. We're not time constrained here so we can re-generated it now.
-
-        // We need to run this inside a clean AsyncLocalStorage snapshot so that the cache
-        // generation cannot read anything from the context we're currently executing which
-        // might include request specific things like cookies() inside a React.cache().
-        // Note: It is important that we await at least once before this because it lets us
-        // pop out of any stack specific contexts as well - aka "Sync" Local Storage.
-
-        stream = await generateCacheEntry(
-          workStore,
-          workUnitStore,
-          clientReferenceManifestSingleton,
-          cacheHandler,
-          serializedCacheKey,
-          encodedArguments,
-          fn
-        )
-      } else {
-        stream = entry.value
-
-        propagateCacheLifeAndTags(workUnitStore, entry)
-
-        if (entry.stale) {
-          // If this is stale, and we're not in a prerender (i.e. this is dynamic render),
-          // then we should warm up the cache with a fresh revalidated entry.
-          const ignoredStream = await generateCacheEntry(
+      let stream: ReadableStream
+
+      const cacheScope: undefined | CacheScopeStore =
+        cacheScopeAsyncLocalStorage.getStore()
+      if (cacheScope) {
+        // String cache key for easier hash mapping.
+        // Note that we're not worried about collisions between string and base64
+        // since the string form will always be JSON which doesn't overlap with base64.
+        const stringCacheKey =
+          typeof serializedCacheKey === 'string'
+            ? serializedCacheKey
+            : Buffer.from(serializedCacheKey).toString('base64')
+        const cachedStream:
+          | undefined
+          | Promise<{
+              value: ReadableStream
+            }> = cacheScope.cache.get(stringCacheKey)
+        if (cachedStream !== undefined) {
+          const entry = await cachedStream
+          // Get a clone out of the cache.
+          const [streamA, streamB] = entry.value.tee()
+          entry.value = streamB
+          stream = streamA
+        } else {
+          // If this is a miss we need to synchronously save an entry so that if we're asked again
+          // for the same entry, we can dedupe the requests. So we split the Promise into two Promises
+          // of streams.
+          const loadedStream = loadCacheEntry(
             workStore,
             workUnitStore,
+            cacheScope,
             clientReferenceManifestSingleton,
             cacheHandler,
             serializedCacheKey,
             encodedArguments,
             fn
           )
-          await ignoredStream.cancel()
+          const streams = teePromiseOfStream(loadedStream)
+          cacheScope.cache.set(
+            stringCacheKey,
+            cacheScopeEntryFromSecondStream(streams)
+          )
```

Comment:
By setting with a Promise rather than awaiting first, we ensure that any new reads get deduped before it is computed.